Source/Core/Common/x64Emitter.cpp

```diff
@@ -886,21 +886,45 @@ void XEmitter::WriteMOVBE(int bits, u8 op, X64Reg reg, const OpArg& arg)
 void XEmitter::MOVBE(int bits, X64Reg dest, const OpArg& src) {WriteMOVBE(bits, 0xF0, dest, src);}
 void XEmitter::MOVBE(int bits, const OpArg& dest, X64Reg src) {WriteMOVBE(bits, 0xF1, src, dest);}
 
-void XEmitter::LoadAndSwap(int size, X64Reg dst, const OpArg& src)
+void XEmitter::LoadAndSwap(int size, X64Reg dst, const OpArg& src, bool sign_extend)
 {
-	if (cpu_info.bMOVBE)
-	{
-		MOVBE(size, dst, src);
-	}
-	else
+	switch (size)
 	{
-		MOV(size, R(dst), src);
-		BSWAP(size, dst);
+	case 8:
+		if (sign_extend)
+			MOVSX(32, 8, dst, src);
+		else
+			MOVZX(32, 8, dst, src);
+		break;
+	case 16:
+		MOVZX(32, 16, dst, src);
+		if (sign_extend)
+		{
+			BSWAP(32, dst);
+			SAR(32, R(dst), Imm8(16));
+		}
+		else
+		{
+			ROL(16, R(dst), Imm8(8));
+		}
+		break;
+	case 32:
+	case 64:
+		if (cpu_info.bMOVBE)
+		{
+			MOVBE(size, dst, src);
+		}
+		else
+		{
+			MOV(size, R(dst), src);
+			BSWAP(size, dst);
+		}
```

Comment:
This should have a break on it as well (higher warning levels will complain about this)